src/Tools/DTO/Tool.php

```diff
@@ -104,32 +118,78 @@ public static function getJsonSchema(): array
                 [
                     'type' => 'object',
                     'properties' => [
-                        'type' => [
+                        self::KEY_TYPE => [
                             'type' => 'string',
                             'const' => ToolTypeEnum::functionDeclarations()->value,
                             'description' => 'The type of tool.',
                         ],
-                        'functionDeclarations' => [
+                        self::KEY_FUNCTION_DECLARATIONS => [
                             'type' => 'array',
                             'items' => FunctionDeclaration::getJsonSchema(),
                             'description' => 'Function declarations.',
                         ],
                     ],
-                    'required' => ['type', 'functionDeclarations'],
+                    'required' => [self::KEY_TYPE, self::KEY_FUNCTION_DECLARATIONS],
                 ],
                 [
                     'type' => 'object',
                     'properties' => [
-                        'type' => [
+                        self::KEY_TYPE => [
                             'type' => 'string',
                             'const' => ToolTypeEnum::webSearch()->value,
                             'description' => 'The type of tool.',
                         ],
-                        'webSearch' => WebSearch::getJsonSchema(),
+                        self::KEY_WEB_SEARCH => WebSearch::getJsonSchema(),
                     ],
-                    'required' => ['type', 'webSearch'],
+                    'required' => [self::KEY_TYPE, self::KEY_WEB_SEARCH],
                 ],
             ],
         ];
     }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @since n.e.x.t
+     *
+     * @return ToolArrayShape
+     */
+    public function toArray(): array
+    {
+        $data = [self::KEY_TYPE => $this->type->value];
+
+        if ($this->type->isFunctionDeclarations() && $this->functionDeclarations !== null) {
+            $data[self::KEY_FUNCTION_DECLARATIONS] = array_map(function (FunctionDeclaration $declaration) {
+                return $declaration->toArray();
+            }, $this->functionDeclarations);
+        } elseif ($this->type->isWebSearch() && $this->webSearch !== null) {
+            $data[self::KEY_WEB_SEARCH] = $this->webSearch->toArray();
+        }
+
+        return $data;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @since n.e.x.t
+     */
+    public static function fromArray(array $array): self
+    {
+        static::validateFromArrayData($array, [self::KEY_TYPE]);
```

Comment:
See my similar note on `MessagePart`, I think we can omit this, since it's inferred from the other data below.